docs/ipdk-security.md

```diff
@@ -0,0 +1,97 @@
+# IPDK Security
+
+## Overview
+
+This document captures all the details related to gRPC ports and secure/insecure modes and certificate management.
+
+## gRPC Ports
+
+IPDK uses a secure-by-default model. The gRPC server will secure ports by default. A user may choose to open insecure ports if needed, but this will be a conscience decision taken by the user at run time.
+
+### Secure-by-default (TLS-mode)
+
+P4CP is launched only with gRPC ports open secured via TLS certificates. The port numbers are:
+
+* 9339 - an IANA-registered port for gNMI and gNOI
+* 9559 - an IANA-registered port for P4RT
+
+### Generating TLS Certificates & installing with script
+
+A script is available to generate and install the certificates in order to establish gRPC secure-mode communication. This setup script uses a pre-configured options and uses OpenSSL to generate the certificate and key files. 
+
+To run the script, which generates certificate and key files and installs to default location:
+
+```bash
+$IPDK_RECIPE/scripts/security/setup_certs_tls_mode.sh
+```
+
+### Generating TLS Certificates & installing manually
+
+The gRPC ports are secured using TLS certificates. A Stratum script and reference configuration files are available to assist in generating the certificates. The reference file uses OpenSSL to generate the keys & certificates. However, users may choose other tools.
+
+The reference file uses a simple PKI where a self-signed RootCA is generated and the same RootCA is used to generate server key & cert and client key & cert. This results in a 1-depth level certificate chain, which will suffice for testing and confirmation, but production systems are encouraged to generate multiple depth levels in order to conform to higher security standards.
+
+The Stratum reference files are available here: https://github.com/stratum/stratum/tree/main/tools/tls
+
+All certificates are in PEM format.
+
+To generate the TLS certificates:
+
+* Review the files to verify if configuration settings are as-desired
+* Run the generate-certs.sh script with following command:
+
+```bash
+COMMON_NAME=<IP> ./generate-certs.sh
+or
+COMMON_NAME=<FQDN> ./generate-certs.sh
+or
+COMMON_NAME=localhost ./generate-certs.sh
+```
+
+* Copy the generated ca.crt, stratum.crt and stratum.key to the server running InfraP4D
+* Copy the generated ca.crt, client.crt and client.key to the gRPC client machine
+
+
+#### Certificate installation
+
+InfraP4D will check for certificate in the following default location:
+
+> /usr/share/stratum/certs/
+
+If alternate location is desired, the location will need to be specified during runtime with the following flags
+
+> -ca_cert_file=[CA certificate file]
+> -server_cert_file=[Server certificate file]
+> -server_key_file=[Server private key file]
+
+For e.g.:
```

Comment:
Just say "For example:".

_e.g._ is an abbreviation for the Latin expression _exempli gratia_. It means "for example". _For e.g._ translates literally as "for for example," which is nonsense.

_e.g._ is normally inserted into a longer sentence as part of a supporting phrase. As a lead-in, you want to use the English phrase by itself.